File: origami_master/lib/models/origami_model.dart
import 'package:flutter/foundation.dart';
import 'origami_step.dart';

@immutable
class OrigamiModel {
  final String id;
  final String name;
  final String description;
  final String category;
  final String difficulty;
  final int estimatedMinutes;
  final String imagePath;
  final List<String> materials;
  final List<OrigamiStep> steps;

  const OrigamiModel({
    required this.id,
    required this.name,
    required this.description,
    required this.category,
    required this.difficulty,
    required this.estimatedMinutes,
    required this.imagePath,
    required this.materials,
    required this.steps,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is OrigamiModel &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          name == other.name &&
          description == other.description &&
          category == other.category &&
          difficulty == other.difficulty &&
          estimatedMinutes == other.estimatedMinutes &&
          imagePath == other.imagePath &&
          listEquals(materials, other.materials) &&
          listEquals(steps, other.steps);

  @override
  int get hashCode =>
      id.hashCode ^
      name.hashCode ^
      description.hashCode ^
      category.hashCode ^
      difficulty.hashCode ^
      estimatedMinutes.hashCode ^
      imagePath.hashCode ^
      materials.hashCode ^
      steps.hashCode;
}

File: origami_master/backend/src/OrigamiMaster.Domain/Entities/OrigamiModel.cs
using System;
using System.Collections.Generic;
using OrigamiMaster.Domain.Enums;

namespace OrigamiMaster.Domain.Entities;

public class OrigamiModel
{
    public Guid Id { get; set; }
    public Guid CategoryId { get; set; }
    public Category Category { get; set; } = null!;
    public string Name { get; set; } = string.Empty;
    public string Slug { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public DifficultyLevel Difficulty { get; set; }
    public string? ThumbnailUrl { get; set; }
    public string? CoverImageUrl { get; set; }
    public int EstimatedMinutes { get; set; }
    public string Materials { get; set; } = string.Empty; // Storing as JSON or comma-separated
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    public ICollection<OrigamiStep> Steps { get; set; } = new List<OrigamiStep>();
    public ICollection<OrigamiTag> OrigamiTags { get; set; } = new List<OrigamiTag>();
    public ICollection<Favorite> Favorites { get; set; } = new List<Favorite>();
    public ICollection<CompletedModel> CompletedModels { get; set; } = new List<CompletedModel>();
    public ICollection<RecentView> RecentViews { get; set; } = new List<RecentView>();
}
